contracts/BillingConnector.sol

```diff
@@ -0,0 +1,265 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.16;
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { IBilling } from "./IBilling.sol";
+import { IBillingConnector } from "./IBillingConnector.sol";
+import { Governed } from "./Governed.sol";
+import { ITokenGateway } from "arb-bridge-peripherals/contracts/tokenbridge/libraries/gateway/ITokenGateway.sol";
+import { Rescuable } from "./Rescuable.sol";
+import { IERC20WithPermit } from "./IERC20WithPermit.sol";
+import { L1ArbitrumMessenger } from "./arbitrum/L1ArbitrumMessenger.sol";
+
+/**
+ * @title Billing Connector Contract
+ * @dev The billing contract allows for Graph Tokens to be added by a user. The
+ * tokens are immediately sent to the Billing contract on L2 (Arbitrum)
+ * through the GRT token bridge.
+ */
+contract BillingConnector is IBillingConnector, Governed, Rescuable, L1ArbitrumMessenger {
+    // -- State --
+    // The contract for interacting with The Graph Token
+    IERC20 private immutable graphToken;
+    // The L1 Token Gateway through which tokens are sent to L2
+    ITokenGateway public l1TokenGateway;
+    // The Billing contract in L2 to which we send tokens
+    address public l2Billing;
+    // The Arbitrum Delayed Inbox address
+    address public inbox;
+
+    /**
+     * @dev Address of the L1 token gateway was updated
+     */
+    event L1TokenGatewayUpdated(address l1TokenGateway);
+    /**
+     * @dev Address of the L2 Billing contract was updated
+     */
+    event L2BillingUpdated(address l2Billing);
+    /**
+     * @dev Address of the Arbitrum Inbox contract was updated
+     */
+    event ArbitrumInboxUpdated(address inbox);
+    /**
+     * @dev Tokens sent to the Billing contract on L2
+     */
+    event TokensSentToL2(address indexed _from, address indexed _to, uint256 _amount);
+    /**
+     * @dev Request sent to the Billing contract on L2 to remove tokens from the balance
+     */
+    event TokensRemovedOnL2(address indexed _from, address indexed _to, uint256 _amount);
```

Comment:
I wonder if we should rename this event to something else given that as described below there is no guarantee that the tx on L2 will succeed. Maybe `RequestTokensRemovedOnL2` or something similar

Reply:
Good point, replaced for `RemovalRequestSentToL2`